tools/find_device.c

```diff
@@ -0,0 +1,183 @@
+	struct udev_device *dev;
+	const char *path;
+
+	/* Create the udev object */
+	udev = udev_new();
+	if (!udev)
+	{
+		fprintf(stderr, "Can't create udev\n");
+		goto out;
+	}
+
+	/* Create a list of the devices with the name "ttyACM*". */
+	enumerate = udev_enumerate_new(udev);
+	if (!enumerate)
+	{
+		fprintf(stderr, "Can't create udev enumerate context.\n");
+		goto udev_out;
+	}
+
+	if (udev_enumerate_add_match_sysname(enumerate, sysname) < 0)
+	{
+		fprintf(stderr, "Failed on adding udev enumerate filter.\n");
+		goto enum_out;
+	}
+
+	if (udev_enumerate_scan_devices(enumerate) < 0)
+	{
+		fprintf(stderr, "Failed on scanning device.\n");
+		goto enum_out;
+	}
+
+
+	/* udev_enumerate_get_list_entry returns a list of entry,
+	but we only need the first entry, because we only cares
+	about their parent which are the same for all. */
+	device = udev_enumerate_get_list_entry(enumerate);
+
+
+	/* Get the filename of the /sys entry for the device
+	 and create a udev_device object (dev) representing it */
+	path = udev_list_entry_get_name(device);
+	if (!path)
+	{
+		printf("No matched device found.\n");
+		goto enum_out;
+	}
+	dev = udev_device_new_from_syspath(udev, path);
+
+	/* In order to get information about the
+	 USB device, get the parent device with the
+	 subsystem/devtype pair of "usb"/"usb_device". This will
+	 be several levels up the tree, but the function will find
+	 it.*/
+	dev = udev_device_get_parent_with_subsystem_devtype(
+		dev,
+		"usb",
+		"usb_device");
+	if (!dev)
+	{
+		fprintf(stderr, "Unable to find parent usb device.");
+		goto dev_out;
+	}
+
+	/* From here, we can call get_sysattr_value() for each file
+	 in the device's /sys entry. The strings passed into these
+	 functions (idProduct, idVendor, serial, etc.) correspond
+	 directly to the files in the directory which represents
+	 the USB device. Note that USB strings are Unicode, UCS2
+	 encoded, but the strings returned from
+	 udev_device_get_sysattr_value() are UTF-8 encoded. */
+	if (!strncmp(udev_device_get_sysattr_value(dev, "manufacturer"),
+				 BL_STR_MANUFACTURER, 9) &&
+		!strncmp(udev_device_get_sysattr_value(dev, "product"),
+				 BL_STR_PRODUCT, 29))
+		return true;
+
+dev_out:
+	udev_device_unref(dev);
+enum_out:
+	/* Free the enumerator object */
+	udev_enumerate_unref(enumerate);
+udev_out:
+	udev_unref(udev);
+out:
+	return false;
+}
+
+/* filter function for /dev nodes match */
+static int dev_match(const struct dirent *entry)
+{
+	return !strncmp(entry->d_name, "ttyACM", 6);
+}
+
+int scan(void)
+{
+	struct dirent **namelist;
+	int n, found = 0;
+
+	n = scandir("/dev/", &namelist, dev_match, NULL);
+	if (n == -1)
+	{
+		fprintf(stderr, "scandir fails\n");
+		exit(EXIT_FAILURE);
+	}
+
+	while (n--)
+	{
+		if (match(namelist[n]->d_name))
+		{
+			/* We are expecting /dev/ttyACM* devices which are not
+			   likely to exceed 99, so max size will be 14 */
+			int ret = snprintf(dev_node, 14, "/dev/%s", namelist[n]->d_name);
```

Comment:
```suggestion
			int ret = snprintf(dev_node, sizeof(dev_node) - 1, "/dev/%s", namelist[n]->d_name);
```

What we really care about is not having a buffer overflow.  Maybe I'd make the `dev_node` buffer 32 bytes too, so there's a bit of head room, if someone has loads of devices or something crazy.

Reply:
+1